### PR TITLE
Change role assignment to a DSL.

### PR DESCRIPTION
**Pull request by Lasering**
Add apply methods. Change get methods to return Option.
Models are now final.
Update dependencies.

### CODE REVIEW

File: src/main/scala/pt/tecnico/dsi/openstack/keystone/services/Domains.scala [hunk 1]
```diff
@@ -28,18 +27,27 @@ final class Domains[F[_]: Sync: Client](baseUri: Uri, authToken: Header)
...
     // A domain name is globally unique across all domains.
     list(Query.fromPairs("name" -> name)).compile.lastOrError
   }
+  /**
+   * Get detailed information about the domain specified by name.
+   *
+   * @param name the domain name
+   * @return a Some of the domain matching the name if it exists. A None otherwise.
+   */
+  def getByName(name: String): F[Option[Domain]] =
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/openstack/keystone/services/Domains.scala:45]: Codacy found an issue: [Do not prefix getters with 'get'](https://app.codacy.com/gh/ist-dsi/scala-keystone-client/pullRequest?prid=6017598)

File: src/main/scala/pt/tecnico/dsi/openstack/keystone/services/UniqueWithinDomain.scala [hunk 2]
```diff
@@ -19,14 +19,24 @@ trait UniqueWithinDomain[F[_], T <: Identifiable] { this: CrudService[F, T, _, _
...
+  
+  /**
+   * Get detailed information about the `T` specified by name and domainId.
+   *
+   * @param name the `T` name
+   * @param domainId the domain id
+   * @return a Some `T` matching the name in a specific domain if it exists. A None otherwise.
+   */
+  def get(name: String, domainId: String): F[Option[T]] =
+    list(Query.fromPairs("name" -> name, "domain_id" -> domainId)).compile.last
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/openstack/keystone/services/UniqueWithinDomain.scala:41]: Codacy found an issue: [Usage of last on collections.](https://app.codacy.com/gh/ist-dsi/scala-keystone-client/pullRequest?prid=6017598)

File: src/main/scala/pt/tecnico/dsi/openstack/keystone/services/Domains.scala [hunk 3]
```diff
@@ -28,18 +27,27 @@ final class Domains[F[_]: Sync: Client](baseUri: Uri, authToken: Header)
...
+    list(Query.fromPairs("name" -> name)).compile.last
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/openstack/keystone/services/Domains.scala:46]: Codacy found an issue: [Usage of last on collections.](https://app.codacy.com/gh/ist-dsi/scala-keystone-client/pullRequest?prid=6017598)

File: src/main/scala/pt/tecnico/dsi/openstack/keystone/services/Domains.scala (hunk 1 above)
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/openstack/keystone/services/Domains.scala:45]: Codacy found an issue: [Do not prefix getters with 'get'](https://app.codacy.com/gh/ist-dsi/scala-keystone-client/pullRequest?prid=6017598)

File: src/main/scala/pt/tecnico/dsi/openstack/keystone/services/UniqueWithinDomain.scala (hunk 2 above)
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/openstack/keystone/services/UniqueWithinDomain.scala:41]: Codacy found an issue: [Usage of last on collections.](https://app.codacy.com/gh/ist-dsi/scala-keystone-client/pullRequest?prid=6017598)

File: src/main/scala/pt/tecnico/dsi/openstack/keystone/services/Domains.scala (hunk 3 above)
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/openstack/keystone/services/Domains.scala:46]: Codacy found an issue: [Usage of last on collections.](https://app.codacy.com/gh/ist-dsi/scala-keystone-client/pullRequest?prid=6017598)

File: src/main/scala/pt/tecnico/dsi/openstack/keystone/services/Domains.scala [hunk 4]
```diff
@@ -28,18 +27,27 @@ final class Domains[F[_]: Sync: Client](baseUri: Uri, authToken: Header)
     )))
 
   /**
-    * Get detailed information about the domain specified by name.
+    * Get detailed information about the domain specified by name, assuming it exists.
     *
     * @param name the domain name
-    * @return the domain matching the name.
+    * @return the domain matching the name. If none exists F will contain an error.
```
Comment (afonsomatos) [inline src/main/scala/pt/tecnico/dsi/openstack/keystone/services/Domains.scala:33]: What do you mean F will contain an error? Shouldn't we wrap the call in an `Option`?

File: src/main/scala/pt/tecnico/dsi/openstack/keystone/services/RoleAssignment.scala [hunk 5]
```diff
@@ -1,69 +1,156 @@
...
+    def from(group: Group): F[Unit] = fromGroup(group.id)
+  }
+  final class Is(roleId: String) {
+    private def assignedTo(subjectType: String, id: String): F[Boolean] =
+      HEAD(uri / subjectType / id / "roles" / roleId, authToken).flatMap(client.expectOption(_)(void)).map(_.isDefined)
+    
+    def assignedToUser(id: String): F[Boolean] = assignedTo("users", id)
+    def assignedToGroup(id: String): F[Boolean] = assignedTo("groups", id)
+    
+    // Dotty union types here would be awesome
```
Comment (afonsomatos) [inline src/main/scala/pt/tecnico/dsi/openstack/keystone/services/RoleAssignment.scala:44]: Why? You would have to discriminate between user and group to call the `assignedTo` with the correct argument.

**Review by afonsomatos — APPROVED** [review]

File: src/main/scala/pt/tecnico/dsi/openstack/keystone/services/RoleAssignment.scala (hunk 5 above)
Comment (Lasering) [reply src/main/scala/pt/tecnico/dsi/openstack/keystone/services/RoleAssignment.scala:44]: If you are using the domain classes (eg: User or group) you can invoke `assignedTo` which is overloaded to differentiate between them. But if you want to invoke `assignedTo` using Ids there is no way to differentiate between an User id and a group id, so we need the `assignedToUser` and `assignedToGroup` functions.

**Pull request merged**